cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/workflow/BasicWorkflowToken.java

```diff
@@ -74,28 +78,30 @@ void setCurrentNode(String nodeName) {
    * @param other the other WorkflowToken to be merged
    */
   void mergeToken(BasicWorkflowToken other) {
-    for (Map.Entry<Scope, Map<String, List<NodeValue>>> entry : other.tokenValueMap.entrySet()) {
-      Map<String, List<NodeValue>> thisTokenValueMapForScope = this.tokenValueMap.get(entry.getKey());
+    synchronized (sharedLock) {
+      for (Map.Entry<Scope, Map<String, List<NodeValue>>> entry : other.tokenValueMap.entrySet()) {
+        Map<String, List<NodeValue>> thisTokenValueMapForScope = this.tokenValueMap.get(entry.getKey());
 
-      for (Map.Entry<String, List<NodeValue>> otherTokenValueMapForScopeEntry : entry.getValue().entrySet()) {
-        if (!thisTokenValueMapForScope.containsKey(otherTokenValueMapForScopeEntry.getKey())) {
-          thisTokenValueMapForScope.put(otherTokenValueMapForScopeEntry.getKey(), Lists.<NodeValue>newArrayList());
-        }
-
-        // Iterate over the list of NodeValue corresponding to the current key.
-        // Only add those NodeValue to the merged token which already do not exist.
+        for (Map.Entry<String, List<NodeValue>> otherTokenValueMapForScopeEntry : entry.getValue().entrySet()) {
+          if (!thisTokenValueMapForScope.containsKey(otherTokenValueMapForScopeEntry.getKey())) {
+            thisTokenValueMapForScope.put(otherTokenValueMapForScopeEntry.getKey(), Lists.<NodeValue>newArrayList());
+          }
 
-        for (NodeValue otherNodeValue : otherTokenValueMapForScopeEntry.getValue()) {
-          boolean otherNodeValueExist = false;
-          for (NodeValue thisNodeValue :
-            thisTokenValueMapForScope.get(otherTokenValueMapForScopeEntry.getKey())) {
-            if (thisNodeValue.equals(otherNodeValue)) {
-              otherNodeValueExist = true;
-              break;
+          // Iterate over the list of NodeValue corresponding to the current key.
+          // Only add those NodeValue to the merged token which already do not exist.
+
+          for (NodeValue otherNodeValue : otherTokenValueMapForScopeEntry.getValue()) {
+            boolean otherNodeValueExist = false;
+            for (NodeValue thisNodeValue :
+              thisTokenValueMapForScope.get(otherTokenValueMapForScopeEntry.getKey())) {
```

Comment:
can simplify by creating a set of ``thisTokenValueMapForScope.get(otherTokenValueMapForScopeEntry.getKey())`` and just add if its not present in the set.